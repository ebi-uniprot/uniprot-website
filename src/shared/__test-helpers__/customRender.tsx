/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
