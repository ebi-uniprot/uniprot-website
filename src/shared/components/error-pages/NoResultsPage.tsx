import { lazy, ReactElement, Suspense } from 'react';
import { Message } from 'franklin-sites';

import ErrorBoundary from '../error-component/ErrorBoundary';
import ErrorPage from './ErrorPage';

import ArtWork from './svgs/no-results-found.img.svg';

const UniProtFooter = lazy(
  () => import(/* webpackChunkName: "footer" */ '../layouts/UniProtFooter')
);

const ErrorMessage = () => (
  <Message level="info">
    <small>
      <h1 className="small">Sorry, no results were found!</h1>
      <span>Please try different criteria</span>
    </small>
  </Message>
);

type Props = {
  children?: ReactElement;
};

const NoResultsPage = ({ children = <ErrorMessage /> }: Props) => (
  <>
    <ErrorPage artwork={<img src={ArtWork} width="400" height="400" alt="" />}>
      {children}
    </ErrorPage>
    <ErrorBoundary>
      <Suspense fallback={null}>
        <UniProtFooter />
      </Suspense>
    </ErrorBoundary>
  </>
);

export default NoResultsPage;
