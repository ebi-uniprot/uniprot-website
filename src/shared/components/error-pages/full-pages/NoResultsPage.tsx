import { lazy, ReactElement, Suspense } from 'react';
import { Message } from 'franklin-sites';

import ErrorBoundary from '../../error-component/ErrorBoundary';
import HTMLHead from '../../HTMLHead';
import ErrorComponent from '../ErrorComponent';

import ArtWork from '../svgs/no-results-found.img.svg';

const UniProtFooter = lazy(
  () => import(/* webpackChunkName: "footer" */ '../../layouts/UniProtFooter')
);

const ErrorMessage = () => (
  <Message level="info">
    <small>
      <h1 className="small">Sorry, no results were found!</h1>
      <span>Please try different criteria</span>
    </small>
  </Message>
);

type Props = {
  children?: ReactElement;
};

const NoResultsPage = ({ children = <ErrorMessage /> }: Props) => (
  <>
    <HTMLHead>
      {/* Don't index error pages */}
      <meta name="robots" content="noindex" />
    </HTMLHead>
    <ErrorComponent
      artwork={<img src={ArtWork} width="400" height="400" alt="" />}
    >
      {children}
    </ErrorComponent>
    <ErrorBoundary>
      <Suspense fallback={null}>
        <UniProtFooter />
      </Suspense>
    </ErrorBoundary>
  </>
);

export default NoResultsPage;
