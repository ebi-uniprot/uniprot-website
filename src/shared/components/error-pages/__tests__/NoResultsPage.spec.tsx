import { screen } from '@testing-library/react';

import customRender from '../../../__test-helpers__/customRender';
import NoResultsPage from '../full-pages/NoResultsPage';

jest.mock('../../error-component/ErrorBoundary', () => ({
  __esModule: true,
  default: () => '{{ ErrorBoundary }}',
}));

describe('NoResultsPage component', () => {
  it('should render', () => {
    const { asFragment } = customRender(<NoResultsPage />);
    expect(screen.getByText(/no results were found/)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
