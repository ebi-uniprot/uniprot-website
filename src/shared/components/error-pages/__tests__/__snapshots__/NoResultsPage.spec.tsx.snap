// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoResultsPage component should render 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <img
      alt=""
      height="400"
      src="test-file-stub"
      width="400"
    />
    <div
      class="message message--info"
      role="status"
    >
      <div
        class="message__side-border"
      />
      <test-file-stub
        classname="message--icon-align-top"
        height="1.125em"
        width="1.125em"
      />
      <div
        class="message__title"
      >
        <small>
          <h1
            class="small"
          >
            Sorry, no results were found!
          </h1>
          <span>
            Please try different criteria
          </span>
        </small>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
